src/xrpld/app/tx/detail/Escrow.cpp

```diff
@@ -384,20 +676,217 @@ EscrowFinish::preclaim(PreclaimContext const& ctx)
+    std::shared_ptr<SLE> const& sleDest,
+    STAmount const& balance,
+    STAmount const& amount,
+    AccountID const& issuer,
+    AccountID const& account,
+    AccountID const& dest,
+    bool const& createAsset,
+    beast::Journal journal)
+{
+    bool const srcIssuer = issuer == account;
+    bool const dstIssuer = issuer == dest;
+
+    // Review Note: We could remove this and just say to use batch to auth the
+    // token first
+    auto const issuanceKey =
+        keylet::mptIssuance(amount.get<MPTIssue>().getMptID());
+    if (!view.exists(keylet::mptoken(issuanceKey.key, dest)) && createAsset &&
+        !dstIssuer)
+    {
+        if (std::uint32_t const ownerCount = {sleDest->at(sfOwnerCount)};
+            balance < view.fees().accountReserve(ownerCount + 1))
+        {
+            JLOG(journal.trace()) << "MPToken does not exist. "
+                                     "Insufficent reserve to create line.";
+            return tecINSUFFICIENT_RESERVE;
+        }
+
+        /*
+        This should have been added to rippleCreditMPT as rippleCreditIOU
+        creates the trustline
+        */
+        auto const mptokenKey = keylet::mptoken(issuanceKey.key, dest);
+        auto const ownerNode = view.dirInsert(
+            keylet::ownerDir(dest), mptokenKey, describeOwnerDir(dest));
+
+        if (!ownerNode)
+            return tecDIR_FULL;
+
+        auto mptoken = std::make_shared<SLE>(mptokenKey);
+        (*mptoken)[sfAccount] = dest;
+        (*mptoken)[sfMPTokenIssuanceID] = amount.get<MPTIssue>().getMptID();
+        (*mptoken)[sfFlags] = 0;
+        (*mptoken)[sfOwnerNode] = *ownerNode;
+        view.insert(mptoken);
+
+        // Update owner count.
+        adjustOwnerCount(view, sleDest, 1, journal);
+    }
+
+    if (!view.exists(keylet::mptoken(issuanceKey.key, dest)) && !dstIssuer)
+        return tecNO_AUTH;
+
+    auto const xferRate = transferRate(view, issuer);
+    // update if issuer rate is less than locked rate
+    if (xferRate < lockedRate)
+        lockedRate = xferRate;
+
+    // Transfer Rate only applies when:
+    // 1. Issuer is not involved in the transfer (srcIssuer or dstIssuer)
+    // 2. The locked rate is different from the parity rate
+
+    auto finalAmt = amount;
+    if ((!srcIssuer && !dstIssuer) && lockedRate != parityRate)
+    {
+        // compute transfer fee, if any
+        auto const xferFee = amount.value() -
+            divideRound(amount, lockedRate, amount.issue(), true);
+        // compute balance to transfer
+        finalAmt = amount.value() - xferFee;
+    }
+
+    return rippleCredit(
+        view,
+        issuer,
+        dest,
+        finalAmt,
+        /*checkIssuer*/ false,
+        journal);
+    return tesSUCCESS;
+}
+
 TER
 EscrowFinish::doApply()
 {
+    PaymentSandbox psb(&ctx_.view());
     auto const k = keylet::escrow(ctx_.tx[sfOwner], ctx_.tx[sfOfferSequence]);
-    auto const slep = ctx_.view().peek(k);
+    auto const slep = psb.peek(k);
```

Comment:
doApply function has become too big. Please consider to move out some part of it to helper functions.